mirgecom/logging_quantities.py

```diff
@@ -0,0 +1,306 @@
+
+
+def logmgr_add_default_discretization_quantities(logmgr: LogManager, discr, dim):
+    """Add default discretization quantities to the logmgr."""
+    for quantity in ["pressure", "temperature"]:
+        for op in ["min", "max", "norm"]:
+            logmgr.add_quantity(DiscretizationBasedQuantity(
+                discr, quantity, op))
+    for quantity in ["mass", "energy"]:
+        for op in ["min", "max", "norm"]:
+            logmgr.add_quantity(DiscretizationBasedQuantity(
+                discr, quantity, op))
+
+    for dim in range(dim):
+        for op in ["min", "max", "norm"]:
+            logmgr.add_quantity(DiscretizationBasedQuantity(
+                discr, "momentum", op, dim=dim))
+
+
+# {{{ Package versions
+
+def add_package_versions(mgr: LogManager, path_to_version_sh: str = None) -> None:
+    """Add the output of the emirge version.sh script to the log.
+
+    Parameters
+    ----------
+    mgr
+        The :class:`logpyle.LogManager` to add the versions to.
+
+    path_to_version_sh
+        Path to emirge's version.sh script. The function will attempt to find this
+        script automatically if this argument is not specified.
+
+    """
+    import subprocess
+    from warnings import warn
+
+    output = None
+
+    # Find emirge's version.sh in any parent directory
+    if path_to_version_sh is None:
+        import pathlib
+        import mirgecom
+
+        p = pathlib.Path(mirgecom.__file__).resolve()
+
+        for d in p.parents:
+            candidate = pathlib.Path(d).joinpath("version.sh")
+            if candidate.is_file():
+                with open(candidate) as f:
+                    if "emirge" in f.read():
+                        path_to_version_sh = str(candidate)
+                        break
+
+    if path_to_version_sh is None:
+        warn("Could not find emirge's version.sh.")
+
+    else:
+        try:
+            output = subprocess.check_output(path_to_version_sh)
+        except OSError as e:
+            warn("Could not record emirge's package versions: " + str(e))
+
+    mgr.set_constant("emirge_package_versions", output)
+
+# }}}
+
+
+# {{{ State handling
+
+def set_sim_state(mgr: LogManager, dim, state, eos) -> None:
+    """Update the simulation state of all :class:`StateConsumer` of the log manager.
+
+    Parameters
+    ----------
+    mgr
+        The :class:`logpyle.LogManager` to set the state of.
+    """
+    for gd_lst in [mgr.before_gather_descriptors,
+            mgr.after_gather_descriptors]:
+        for gd in gd_lst:
+            if isinstance(gd.quantity, StateConsumer):
+                gd.quantity.set_sim_state(dim, state, eos)
+
+
+class StateConsumer:
+    """Base class for quantities that require a state for logging."""
+
+    def __init__(self):
+        self.state_vars = None
+
+    def set_sim_state(self, dim, state, eos) -> None:
+        """Update the state vector of the object."""
+        self.state_vars = extract_state_vars(dim, state, eos)
```

Comment:
Related to previous comment about `eos.dependent_vars` - we just need to be careful because this can get expensive.